### DISCUSSION TITLE
To extract all tables in PDF

### DISCUSSION THREAD

**Discussion by wingbroken** [pymupdf/PyMuPDF#1128] [Ideas]
I know there are some examples that can extract table contents from PDF with Pymupdf. But seems they require pre-analysis of specific doc to adress the table. Any generic way we can extract all tables in PDF like pdfplumber.extract_table() does via some prior configuration? If not, consider to enhance Pymupdf? I like PYMUPDF's fast speed and more granular funtions.

**Comment by JorjMcKie** [comment]
I understand your motivation.
There are at least two or three considerations that let me turn down this enhancement proposal:
1. When exporting spreadsheet documents to PDFs (like from Excel or LibreOffice), the result **_will not conform_** to the table formalism as defined in the PDF spec. So an important (if not critical) condition of implementing this is not met.
2. Analyzing a table that is indeed specified using PDF spec, is a very, **_very_** complex undertaking.
3. Detecting and then analyzing tables, which are basically defined as text, is even more challenging. I have seen a few heuristics approaches which use pretty heavy calibers. Clearly beyond of what I consider PyMuPDF's scope.

Taking the three together, there is no business case for me doing this.
As PyMuPDF does offer ways accessing low-level structures in PDFs using Python, nothing stands in the way for someone who wants to try it.

**Reply by ram02z** [reply]
>     3. Detecting and then analyzing tables, which are basically defined as text, is even more challenging. I have seen a few heuristics approaches which use pretty heavy calibers

Could you please point me to one of the approaches you mention?

**Reply by JorjMcKie** [reply]
For example here: https://medium.com/mlearning-ai/extracting-tabular-data-from-pdf-using-deep-learning-table-detection-81242cbad6d

But also this: https://www.pdftron.com/pdf-tools/pdf-table-recognition/

**Reply by ram02z** [reply]
> For example here: https://medium.com/mlearning-ai/extracting-tabular-data-from-pdf-using-deep-learning-table-detection-81242cbad6d
> 
> But also this: https://www.pdftron.com/pdf-tools/pdf-table-recognition/

Thank you!

**Reply by ram02z** [reply]
Sorry for the followup question, but do you know of any approaches that didn't use AI?
